flink-table/flink-table-common/src/test/java/org/apache/flink/table/connector/source/TestManagedTableSource.java

```diff
@@ -0,0 +1,69 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.flink.table.connector.source;
+
+import org.apache.flink.table.connector.ChangelogMode;
+import org.apache.flink.table.factories.DynamicTableFactory;
+
+/** Managed {@link DynamicTableSource} for testing. */
+public class TestManagedTableSource implements ScanTableSource {
+
+    private final DynamicTableFactory.Context context;
+    private final CompactPartitions partitions;
+    private final ChangelogMode changelogMode;
+
+    public TestManagedTableSource(
+            DynamicTableFactory.Context context,
+            CompactPartitions partitions,
+            ChangelogMode changelogMode) {
+        this.context = context;
+        this.partitions = partitions;
+        this.changelogMode = changelogMode;
+    }
+
+    @Override
+    public ChangelogMode getChangelogMode() {
+        return changelogMode;
+    }
+
+    @Override
+    public ScanRuntimeProvider getScanRuntimeProvider(ScanContext runtimeProviderContext) {
+        return SourceProvider.of(new TestManagedSource(partitions));
+    }
+
+    @Override
+    public DynamicTableSource copy() {
+        return new TestManagedTableSource(context, partitions, changelogMode);
+    }
+
+    @Override
+    public String asSummaryString() {
+        return "TestManagedTableSource";
+    }
+
+    @Override
+    public boolean equals(Object o) {
+        return super.equals(o);
```

Comment:
equals and hashcode?

Reply:
> equals and hashcode?

Aha, I just copied from the previous commit. I will change it.